src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs

```diff
@@ -521,7 +525,13 @@ internal override void AbortRead(long errorCode)
 
         internal override void AbortWrite(long errorCode)
         {
-            ThrowIfDisposed();
+            if (_disposed == 1)
+            {
+                // Dispose already triggered graceful shutdown
+                // It is unsafe to try to trigger abortive shutdown now, because final event arriving after Dispose releases SafeHandle
+                // so if it arrives after our check but before we call msquic, me might end up with access violation
```

Comment:
Would it help if we send the abort only if the `SendState` was < `Aborted`? Few lines bellow we test the state, we could as well use similar pattern we do elsewhere with `shouldAbort`, set it to true only if we change the state and send the Abort only in such cases.

If this is something we could do, it might also help with the concern you raised in https://github.com/dotnet/runtime/pull/57156, since it would provide the "locking" security via the `SendState` property.

Reply:
Discussed offline that the suggestion is good, but not directly related to the change, so it should be done in a separate PR